### DISCUSSION TITLE
Async method call with mongock

### DISCUSSION THREAD

**Discussion by manishpurohit92** [flamingock/mongock#45] [Q&A] [answered]
Does mongock support Async call? I mean, Is it possible to run long-running migration (30 min) in the background. Ex method returning CompletableFuture<Void>.
 Or it will release the lock and not able to finish the migration.

**Comment by dieppa** [comment, marked as answer]
Hello.

can you explain me a bit more about your scenario? If you are running a 30 mins migration and don't mind to start the application/service while it's not finished yet. Why don't run you migration separately?

**Reply by manishpurohit92** [reply]
Migration runs when the event comes in. (for example when the first request comes). We are not running this on application start-up. and it's taking more than 30min.

**Reply by manishpurohit92** [reply]
yes run separately, but can't we run asynchronously? asking because it can take more than 1 hour, and what if it will be not able to acquire the lock for a long amount of time.

**Reply by dieppa** [reply]
OK, I would like to understand a bit better your scenario, so can be in the FAQ for future users:

**- What happens is two similar requests come? They will try to run the same migration? I guess only one should able to run it, right?**

OK, if it's what I think it is. It would work. The lock mechanism works in such a way that is acquired initially for a period. If it finishes before the the acquired time, the lock is released. In the worst case, Mongock process is abruptly interrupted and the lock is eventually expired, so another Mongock process can take it.

On the other hand, if the mongock process is not still finished by the time the lock expires, it's refreshed(it obviously doesn't wait until the last nanosecond, too risky, it does it with some margin to guarantee  no conflicts). This will happens "for ever" until either the Mongock process is finished, an exception happens or the JVM is interrupted.

So having this into account, I wouldn't configure Mongock to retain the lock for too long as, in case of non-graceful failure, it will prevent your application from starting any migration during that time the lock is "reserved" (By default 3 mins). If you are sure your migrations take that long and you are confident they are stable and won't break, leaving the lock unreleased, , you can increase that  value a bit, so Mongock is not refreshing the lock so often(By default would do it roughly 10 times for 30-mins-migration and 20 for 60-mins-migration), but still that is not the end of the world in a general case. 

I hope that resolves your question and Mongock is the answer to your needs :)

**PD: As said, if you can provide more information about your scenario, so it can be helpful for future potential users with the same issue, it would be great.**

**Comment by manishpurohit92** [comment]
More Details:
We want to update 5M documents for a user and we have many users with different document sizes. So doing this heavy update, we can't do it in sync fashion. So We are doing the async call.   

after the migration is done. ( Changock has finished) .
It is printing this "Changock trying to refresh the lock"  7-8 times. any reason. why it is trying to acquire the lock.

Let me know, if my information is still not make sense.

**Reply by dieppa** [reply]
> after the migration is done. ( Changock has finished) .
It is printing this "Changock trying to refresh the lock" 7-8 times. any reason. why it is trying to acquire the lock.

As I mentioned, the lock is initially acquired for some time(if you haven't specified anything different, the default value is 3 mins). After that time is consumed(actually +-75% of that time), Mongock refreshes the lock for the same  period again. And then, once it's consumed, it refreshes it again... and again and again...Until either the migration is finished or an exception occurs. 

So, if you have a migration that takes 30 mins and you reserve the lock for 3 minutes, Mongock will refresh it 9 times. 

In summary, those messages about refreshing the lock are absolutely normal 😉

**Reply by manishpurohit92** [reply]
But I am running this asynchronously, so migration is already done. So MongoDB update query should run in the background and mongock should release the lock.

**Reply by dieppa** [reply]
Sorry, I don't get that. Mongock will release the lock as soon as the Mongock process is finished. And then another Mongock process can start. 

Please note that if a Mongock migration is in progress and a new request, therefor another migration needs to be executed, it won't start until the the on in progress finished. That's the goal of the lock, synchronise migration so only one is guaranteed to be executed at a time.

**Reply by manishpurohit92** [reply]
@dieppa wants to check with you. why I am not able to run mongock migration asynchronously.
Example
Code inside async is to update 100k entries.  and Lock is with the default setting. I tested it's not updating any entries.
when I am removing async part, it's working.  I don't want to run this long process synchronously.  

Ex:
    @ChangeSet(order = "001", id = "test async ", author = "me")
    public CompletableFuture<Void> test(
            MongoTemplate mongockTemplate
    ) {
        return CompletableFuture.runAsync(() ->
                testAsync(
                        mongockTemplate
                )
        );
    }
private void testAsync(MongoTemplate mongockTemplate) {
// code to update 5M entry
}

**Reply by dieppa** [reply]
Hello, you asked about the lock. Nothing about to return a Callable from a changeSet. I didn't know you were doing it that way. Of course that won't work. Mongock doesn't do anything with the returned value of a changeSet. Actually should return void. When you mentioned to run it asynchronously with a callable, I thought you were wrapping the `runner.execute()` in a callable. 

Before anything, you should have clear that Mongock(at least currently) is not intended to be used in that way.

Having that into account, I can suggest you some solution, but I need to have the whole picture, maybe a sequence.
Is this right? You start your server, your Mongock bean is created, but the migration is not run yet. Then when a request comes, you call the method `execute()` in the MongockRunner and expected to call that changeSet asynchronously.

Assuming that is your sequence.
1) What happens when another request comes? It will try to run the same changeSet and your expectation is that waits, asynchronously,  for the lock and when is available execute the migration..again?
2) Or once that changeSet is executed, shouldn't be executed again?
3)Why using Mongock? Why not a normal async method ?


To be honest I am totally confused. Don't understand what you are trying to do, why you expected the callable to work and not sure if Mongock is the best tool for you in this scenario.

**Reply by manishpurohit92** [reply]
@dieppa 
Sorry for all the confusion.  
Sequence is : 
1 and 2>
When a request comes, I will create mongockbean, and call the method execute() in the MongockRunner and expected to call that changeSet asynchronously.
once that changeSet is executed, It shouldn't be executed again. correct.
and yes I am not expecting the return value (void) by thread. Earlier code has a typo. Again putting this again.


   @Changeset(order = "001", id = "test async ", author = "me")
    public void test(
            MongoTemplate mongockTemplate
    ) {
         CompletableFuture.runAsync(() ->
                testAsync(
                        mongockTemplate
                )
        );
    }

    private void testAsync(MongoTemplate mongockTemplate) {
// code to update 5M entry
    }

3> I would need to update multiple documents for multiple environments. that's why using mongock.  I can write the script but that would take a long time.
So checking if I can achieve this with mongock.

**Reply by dieppa** [reply]
Ok, so, I see that you have a migration that only runs once. OK, assuming you have that part under control, I suggest you to remove the ComplatableFuture from the changeLog/ChangeSet, they should be clean. 

```java
@Changeset(order = "001", id = "test async ", author = "me")
public void test(MongoTemplate mongockTemplate) {
   testAsync(mongockTemplate);
}

private void testAsync(MongoTemplate mongockTemplate) {
// code to update 5M entry
}
```

Then you can wrap the call to `MongockRunner.execute()` in a async function, like you are doing. Something like this:
```java
CompletableFuture.runAsync(mongockRunner::execute);
```

If you are using Spring, you could also do something like this
```java
@Async
public void runMigration() {
   mongockRunner.execute()
}
```

Although Mongock is not intended to be used in this way, I believe that this could work for you.

Let me know.

**Comment by eleco** [comment]
hello @dieppa  - is there any code sample that would show how to run a changeset asynchronously ? using spring boot and Java preferably.

In particular not sure how to get hold of a mongockRunner instance..thanks.

**Reply by dieppa** [reply]
Hello @eleco ,

what do you mean by " run a changeSet asynchronously"? Do you mean running the changeUnit(old changeSet) independently from the application, with the CLI or running the changeUnits within the application, but in parallel with the application's startup?

**Comment by eleco** [comment]
I have a changeset to create an index. Running that change set takes a long time. Is there a way to run it asynchronously, so that the creation of the index does not block the application from running ? 

 **running the changeUnits within the application, but in parallel with the application's startup** -> that's what I want :)

**Comment by dieppa** [comment]
Ok, I got what you mean now. 

It's a bit tricky because a changeUnit shouldn't be async. The changeUnits should be executed one after the other. So you need to run the entire migration asynchronously.

Please take a look to [this discussion](https://github.com/mongock/mongock/discussions/402)  and let me know if it's what you are looking for

**Reply by eleco** [reply]
that's great - just what I was looking for - thanks.
